rasa/core/policies/rule_policy.py

```diff
@@ -74,6 +80,154 @@ def _create_feature_key(self, states: List[Dict]) -> Text:
 
         return feature_str
 
+    @staticmethod
+    def _get_active_form_name(state: Dict[Text, float]) -> Optional[Text]:
+        found_forms = [
+            state_name[len(ACTIVE_FORM_PREFIX) :]
+            for state_name, prob in state.items()
+            if ACTIVE_FORM_PREFIX in state_name
+            and state_name != ACTIVE_FORM_PREFIX + "None"
+            and prob > 0
+        ]
+        # by construction there is only one active form
+        return found_forms[0] if found_forms else None
+
+    @staticmethod
+    def _prev_action_listen_in_state(state: Dict[Text, float]) -> bool:
+        return any(
+            PREV_PREFIX + ACTION_LISTEN_NAME in state_name and prob > 0
+            for state_name, prob in state.items()
+        )
+
+    @staticmethod
+    def _modified_states(
+        states: List[Dict[Text, float]]
+    ) -> List[Optional[Dict[Text, float]]]:
+        """Modify the states to
+            - capture previous meaningful action before action_listen
+            - ignore previous intent
+        """
+        if states[0] is None:
+            action_before_listen = None
+        else:
+            action_before_listen = {
+                state_name: prob
+                for state_name, prob in states[0].items()
+                if PREV_PREFIX in state_name and prob > 0
+            }
+        # add `prev_...` to show that it should not be a first turn
+        indicator = PREV_PREFIX + RULE_SNIPPET_ACTION_NAME
+        if indicator not in action_before_listen.keys():
+            return [{indicator: 1}, action_before_listen, states[-1]]
+        return [action_before_listen, states[-1]]
+
+    @staticmethod
+    def _clean_feature_keys(lookup: Optional[Dict]) -> Optional[Dict]:
+        # remove action_listens that were added after conditions
+        updated_lookup = lookup.copy()
+        for key in lookup.keys():
```

Comment:
When you also get the `predicted_action` right away here you don't need to retrieve it again in `lookup[key]` (line 131). 
```suggestion
        for states, predicted_action in lookup.items():
```